vltt/assignment3/solution.cpp

```diff
@@ -0,0 +1,197 @@
+  }
+  
+
+template <class T>
+inline std::ostream& operator<<(std::ostream& os, const std::set<T>& s) {
+    os << "{";
+    for (auto &it : s) {
+        os << it;
+        std::cout << ", " ;
+    }
+    os << "}";
+    return os;
+}
+
+class Dictionary {
+private:
+    class Trie {
+    private:
+        class Node {
+        private:
+            std::map<char, Node*> go;
+        public:
+            bool terminal = false;
+            void add(const char &letter) {
+                go[letter] = new Node;
+            }
+            bool isHere(const char &letter) {
+                return go.find(letter) != go.end();
+            }
+            Node *next(const char &letter) {
+                if (go.find(letter) == go.end()) return nullptr;
+                return go[letter];
+            }
+        };
+        Node *start;
+    public:
+        Trie() {
+            start = new Node;
+        }
+        //add new word to the Trie
+        void add(const std::string &word) {
+            if (word.empty()) return;
+            Node *cur = start;
+            for (int i = 0; i < word.length(); i++) {
+                char letter = word[i];
+                if (cur->isHere(letter) == false) {
+                    cur->add(letter);
+                }
+                cur = cur->next(letter);
+                if (i == int(word.length()) - 1) {
+                    cur->terminal = true;
+                }
+            }
+        }
+        //determines whether there is such a prefix in the Trie
+        //return: -1 - prefix, 0 - nothing, 1 - word
+        int find_str(const std::string &str) {
+            if (str.empty()) return 0;
+            Node *cur = start;
+            for (int i = 0; i < str.length(); i++) {
+                if (cur == nullptr) break;
+                char letter = str[i];
+                cur = cur->next(letter);
+                if (i == int(str.length()) - 1) {
+                    if (cur == nullptr) return 0;
+                    if (cur->terminal == true) return 1;
+                    else return -1;
+                }
+            }
+            return 0;
+        }
+    };
+    Trie prefixes;
+public:
+    Dictionary() {}
+    Dictionary(std::vector<std::string> all_words) {
+        for (auto &word : all_words) {
+            add(word);
+        }
+    }
+    //add new word to the Dictionary
+    void add(const std::string &word) {
+        prefixes.add(word);
+    }
+    bool isWord(const std::string &cur) {
+        return prefixes.find_str(cur) == 1;
+    }
+    bool isPrefix(const std::string &cur) {
+        return prefixes.find_str(cur);
+    }
+};
+
+void passage_through_the_grid(const std::vector<std::vector<char> > &grid, Dictionary &dict, std::string &cur, int i_gr, int j_gr, std::set<std::string> &ans, std::set<std::pair<int, int> > &used) {
+    const std::vector<int> id_x = { 0, 1, -1 };
+    const std::vector<int> id_y = { 0, 1, -1 };
+    for (int i = 0; i < 3; i++) {
```

Comment:
I would probably use `int i = 0, i < id_x.side()` and `int j = 0, j < id_y.side()` just to be super safe :)